aspnetcore/security/authentication/mfa.md

```diff
@@ -0,0 +1,562 @@
+                throw new ArgumentNullException(nameof(requirement));
+
+            var amrClaim = context.User.Claims.FirstOrDefault(t => t.Type == "amr");
+
+            if (amrClaim != null && amrClaim.Value == Amr.Mfa)
+            {
+                context.Succeed(requirement);
+            }
+
+            return Task.CompletedTask;
+        }
+    }
+}
+```
+
+In the `Startup.ConfigureServices` method, the `AddOpenIdConnect` method is used as the default challenge scheme. The authorization handler, which is used to check the `amr` claim, is added to the Inversion of Control container. A policy is then created which adds the `RequireMfa` requirement.
+
+```csharp
+public void ConfigureServices(IServiceCollection services)
+{
+	services.ConfigureApplicationCookie(options =>
+	{
+		options.Cookie.SecurePolicy = Microsoft.AspNetCore.Http.CookieSecurePolicy.Always;
+	});
+
+	services.AddSingleton<IAuthorizationHandler, RequireMfaHandler>();
+
+	services.AddAuthentication(options =>
+	{
+		options.DefaultScheme = CookieAuthenticationDefaults.AuthenticationScheme;
+		options.DefaultChallengeScheme = OpenIdConnectDefaults.AuthenticationScheme;
+	})
+	.AddCookie()
+	.AddOpenIdConnect(options =>
+	{
+		options.SignInScheme = "Cookies";
+		options.Authority = "https://localhost:44352";
+		options.RequireHttpsMetadata = true;
+		options.ClientId = "AspNetCoreRequireMfaOidc";
+		options.ClientSecret = "AspNetCoreRequireMfaOidcSecret";
+		options.ResponseType = "code id_token";
+		options.Scope.Add("profile");
+		options.Scope.Add("offline_access");
+		options.SaveTokens = true;
+	});
+
+	services.AddAuthorization(options =>
+	{
+		options.AddPolicy("RequireMfa", policyIsAdminRequirement =>
+		{
+			policyIsAdminRequirement.Requirements.Add(new RequireMfa());
+		});
+	});
+
+	services.AddRazorPages();
+}
+```
+
+This policy is then used in the Razor page as required. The policy could be added globally for the entire app as well.
+
+```csharp
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Mvc.RazorPages;
+using Microsoft.Extensions.Logging;
+
+namespace AspNetCoreRequireMfaOidc.Pages
+{
+    [Authorize(Policy= "RequireMfa")]
+    public class IndexModel : PageModel
+    {
+        private readonly ILogger<IndexModel> _logger;
+
+        public IndexModel(ILogger<IndexModel> logger)
+        {
+            _logger = logger;
+        }
+
+        public void OnGet()
+        {
+
+        }
+    }
+}
+```
+
+If the user authenticates without MFA, the `amr` claim will probably have a `pwd` value. The request won't be authorized to access the page. Using the default values, the user will be redirected to the account/AccessDenied page. This can be changed or you can implement your own custom logic here. In this example, a link is added so that the valid user can set up MFA for his or her account.
```

Comment:
```suggestion
If the user authenticates without MFA, the `amr` claim will probably have a `pwd` value. The request won't be authorized to access the page. Using the default values, the user will be redirected to the *account/AccessDenied* page. This can be changed or you can implement your own custom logic here. In this example, a link is added so that the valid user can set up MFA for his or her account.
```

Reply:
Ah there's the special claim :)

Reply:
fixed the texted here